### PR TITLE
Add gcsweb

### PR DESCRIPTION
**Pull request by thockin**
gcsweb is a tiny web frontend to GCS browsing that DOES NOT REQUIRE A GOOGLE LOGIN.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/test-infra/728)

### CODE REVIEW

File: gcsweb/cmd/gcsweb/gcsweb.go [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+		logger.Printf("GET %s: %s", url, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "http.Get: %v", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	logger.Printf("GET %s: %s", url, resp.Status)
+
+	if resp.StatusCode != http.StatusOK {
```
Comment (rmmh) [inline gcsweb/cmd/gcsweb/gcsweb.go]: The GCS documentation suggests retrying 500s with exponential backoff. Why not do that here?

**Comment by rmmh** [comment]
Can you modify this to use `html/template` instead of the Fprintfs? It's generally more readable and does autoescaping.

You can keep the templates inline for simplicity, as globals defined using [Must](
https://golang.org/pkg/html/template/#Must)
```go
var tmlpIndex = template.Must(template.New("index").Parse(`
<!DOCTYPE html>
<body>
{{ range .Entries }}
<p>{{ . }}
{{ end }}
`))

```

**Comment by thockin** [comment]
There is so little html...

I will look at it, but Might ask for a deferment?

**Comment by spxtr** [comment]
Spiffy. I don't think we need to include another copy of license here, since it's the same for all test-infra projects.

**Comment by thockin** [comment]
LICENSE removed

Converted to templates.  Still more functional than templated, but that's a bigger change that can be deferred to a little later?

File: gcsweb/cmd/gcsweb/gcsweb.go (hunk 1 above)
Comment (thockin) [inline gcsweb/cmd/gcsweb/gcsweb.go]: trying to be dead simple.  Can revisit if it is a problem?

**Comment by thockin** [comment]
RC1 is running at http://gcsweb.k8s.io/gcs/kubernetes-jenkins/

**Comment by ixdy** [comment]
where's the config for gcsweb.k8s.io?

File: gcsweb/Makefile [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+# limitations under the License.
+
+# The binary to build (just the basename).
+BIN := gcsweb
+
+# This repo's root import path (under GOPATH).
+PKG := k8s.io/test-infra/gcsweb
+
+# Where to push the docker image.
+REGISTRY ?= gcr.io/google_containers
```
Comment (thockin) [inline gcsweb/Makefile]: We can subdir undir gcr.io/google_containers, too.  Though this is a pretty generic container - I would not at all be surprised if people in the wild wanted to use it.

File: gcsweb/cmd/gcsweb/gcsweb.go [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+		fmt.Println(version.VERSION)
+		os.Exit(0)
+	}
+
+	log.Printf("starting")
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	// Canonicalize allowed buckets.
+	for i := range allowedBuckets {
+		allowedBuckets[i] = joinPath(gcsPath, allowedBuckets[i])
```
Comment (thockin) [inline gcsweb/cmd/gcsweb/gcsweb.go]: will do

File: gcsweb/Makefile [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	@gcloud docker push $(IMAGE):$(VERSION)
+	@docker images -q $(IMAGE):$(VERSION) > $@
+
+push-name:
+	@echo "pushed: $(IMAGE):$(VERSION)"
+
+version:
+	@echo $(VERSION)
+
+test: build-dirs
```
Comment (thockin) [inline gcsweb/Makefile]: will do

File: gcsweb/cmd/gcsweb/gcsweb.go (hunk 3 above)
Comment (thockin) [inline gcsweb/cmd/gcsweb/gcsweb.go]: done

File: gcsweb/cmd/gcsweb/gcsweb.go [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+	http.HandleFunc("/healthz", healthzRequest)
+	http.HandleFunc("/", otherRequest)
+	log.Printf("serving on port %d", *flPort)
+	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", *flPort), nil))
+}
+
+func healthzRequest(w http.ResponseWriter, r *http.Request) {
+	logger := newTxnLogger(r)
+	logger.Printf("GET /healthz")
+	fmt.Fprintf(w, "OK")
```
Comment (thockin) [inline gcsweb/cmd/gcsweb/gcsweb.go]: Is it unusual?

File: gcsweb/cmd/gcsweb/gcsweb.go [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+	// e.g. "/gcs/bucket/path/to/object" -> "/bucket/path/to/object"
+	path := strings.TrimPrefix(r.URL.Path, gcsPath)
+	// e.g. "/bucket/path/to/object" -> ["bucket", "path/to/object"]
+	bucket, object := splitBucketObject(path)
+
+	url := joinPath(gcsBaseURL, bucket)
+	url += "?delimiter=/"
+
+	if object != "" {
+		// Adding the last slash forces the server to give me a clue about
```
Comment (thockin) [inline gcsweb/cmd/gcsweb/gcsweb.go]: ACK.  This is a very unpleasant API

File: gcsweb/cmd/gcsweb/gcsweb.go [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+	Contents       []Record `xml:"Contents"`
+	CommonPrefixes []Prefix `xml:"CommonPrefixes"`
+}
+
+func htmlPageHeader(out io.Writer, name string) error {
+	const tmpl = ` 
+	    <!doctype html>
+    	<html>
+    	<head>
+    	    <link rel="stylesheet" href="http://yui.yahooapis.com/pure/0.6.0/pure-min.css">
```
Comment (thockin) [inline gcsweb/cmd/gcsweb/gcsweb.go]: 404

**Comment by thockin** [comment]
Config is https://github.com/kubernetes/k8s.io/pull/16

**Comment by thockin** [comment]
pushed

File: gcsweb/cmd/gcsweb/gcsweb.go (hunk 7 above)
Comment (ixdy) [inline gcsweb/cmd/gcsweb/gcsweb.go]: huh? that link works for me - https://yui-s.yahooapis.com/pure/0.6.0/pure-min.css

Note the hostname changed (adding a -s) in addition to making it https.

File: gcsweb/cmd/gcsweb/gcsweb.go (hunk 5 above)
Comment (ixdy) [inline gcsweb/cmd/gcsweb/gcsweb.go]: I guess I'm just used to internal healthchecks which are lowercase.

**Comment by rmmh** [comment]
Thanks for the template change! Reparsing the template every time isn't great for efficiency, but probably doesn't matter.

Can you make it redirect all unhandled buckets to the GCS storage browser? Otherwise external services will have to decide whether gcsweb can handle displaying a given path.

**Comment by rmmh** [comment]
LGTM overall.

File: gcsweb/cmd/gcsweb/gcsweb.go (hunk 5 above)
Comment (thockin) [inline gcsweb/cmd/gcsweb/gcsweb.go]: fixed

**Comment by thockin** [comment]
Fixed re-parsing.

I looked at a redirect, bvut that suffers the same problems as before - I don't know if a URL is a file or a directory, and no one URl works for both (except the gcsweb URL introduced here).

**Comment by thockin** [comment]
PTAL

**Comment by rmmh** [comment]
You could use the heuristic of "trailing slash == directory". I'd like to make Gubernator point at gcsweb for all artifacts links (directory browsing), without having to validate that gcsweb will handle it properly.

Linking to gcsweb for a particular file is pretty pointless, since it just redirects to storage.googleapis.com.

**Comment by thockin** [comment]
Can we merge this and you can send a patch?  I am out of bandwidth.

**Comment by rmmh** [comment]
Ok, LGTM.

File: .travis.yml [hunk 8]
```diff
@@ -10,6 +10,7 @@ env:
 cache: pip
 services:
     - docker
+sudo: required
```
Comment (ixdy) [inline .travis.yml]: why do we need sudo?

**Comment by thockin** [comment]
The `make` for gcsweb uses docker.

**Comment by thockin** [comment]
Is this repo bot-driven yet?  Or manual merge?

**Comment by ixdy** [comment]
`jenkins/diff-job-config-patch.sh` uses docker too, and it's been working without issue.

no bot here - it's all manual.

**Comment by thockin** [comment]
somebody merge me so I can cut a release properly :)

**Pull request merged**